File: src/HpBar.lua
local HpBar = class("HpBar", function() return cc.Node:create() end)

local hmargin = 1
local vmargin = 1

local backgroudColor = cc.c4f(0, 0, 0, 1)
local greenColor = cc.c4f(0, 1, 47/255.0, 1)
local yellowColor = cc.c4f(1, 1, 0, 1)
local redColor = cc.c4f(1, 0, 0, 1)

local backgroudTextureName = "HpBar_backgroundTextueName"
local greenTextureName = "HpBar_greenTextureName"
local yellowTextureName = "HpBar_yellowTextureName"
local redTextureName = "HpBar_redTextureName"

local yellowPercentage = 50
local redPercentage = 20

function HpBar:ctor(...)
    local arg = {...}
    local size = arg[2]
    assert(#arg == 2, "invalid arg count: " .. #arg .. ", require 2")
    local innerSize = cc.size(size.width - 2 * hmargin, size.height - 2 * vmargin)
    self:setContentSize(size)
    self:setAnchorPoint(cc.p(0.5, 0.5))

    self._bgProgressTimer = self:_addProgressTimer(backgroudColor, size, backgroudTextureName)
    self._redProgressTimer = self:_addProgressTimer(redColor, innerSize, redTextureName)
    self._yellowProgressTimer = self:_addProgressTimer(yellowColor, innerSize, yellowTextureName)
    self._greenProgressTimer = self:_addProgressTimer(greenColor, innerSize, greenTextureName)
    self:setPercentage(1)
end

function HpBar:setPercentage(percentage)
    local p = self:_normalizePercentage(percentage)
    local pt = self:_currentProgressTimer(p)
    pt:setPercentage(p)
end

function HpBar:_currentProgressTimer(percentage)
    if (percentage <= redPercentage) then
        self._greenProgressTimer:setVisible(false)
        self._yellowProgressTimer:setVisible(false)
        self._redProgressTimer:setVisible(true)
        return self._redProgressTimer
    elseif (percentage <= yellowPercentage) then
        self._greenProgressTimer:setVisible(false)
        self._yellowProgressTimer:setVisible(true)
        self._redProgressTimer:setVisible(false)
        return self._yellowProgressTimer
    else
        self._greenProgressTimer:setVisible(true)
        self._yellowProgressTimer:setVisible(false)
        self._redProgressTimer:setVisible(false)
        return self._greenProgressTimer
    end
end

-- private

function HpBar:_normalizePercentage(percentage)
    percentage = percentage * 100
    return percentage > 100 and 100 or (percentage > 0 and percentage or 0)
end

function HpBar:_addProgressTimer(color, size, textureName)
    local textureCache = cc.Director:getInstance():getTextureCache()
    local tex = textureCache:getTextureForKey(textureName)
    if (tex == nil) then
        tex = G.textureWithColor(color, cc.size(1, 1))
        textureCache:addTexture(tex, textureName)
    end
    local sprite = cc.Sprite:createWithTexture(tex)
    local pt = cc.ProgressTimer:create(sprite)
    pt:setType(cc.PROGRESS_TIMER_TYPE_BAR)
    pt:setMidpoint(cc.p(0, 0.5))
    pt:setBarChangeRate(cc.p(1, 0))
    pt:setPercentage(100)
    pt:setNormalizedPosition(cc.p(0.5, 0.5))
    pt:setScale(size.width, size.height)
    self:addChild(pt)
    return pt
end

return HpBar

File: frameworks/runtime-src/Classes/unit/HpBar.cpp
//
//  HpBar.cpp
//  Boids
//
//  Created by Yanjie Chen on 3/16/15.
//
//

#include "HpBar.h"

using namespace cocos2d;

#define BACKGROUND_TEXTURE_NAME "HpBar_backgroundTextureName"
#define GREEN_TEXTUER_NAME "HpBar_greenTextureName"
#define YELLOW_TEXTURE_NAME "HpBar_yellowTextureName"
#define RED_TEXTURE_NAME "HpBar_redTextureName"

#define RED_PERCENTAGE 20.0
#define YELLOW_PERCENTAGE 50.0

#define H_MARGIN 1.0f
#define V_MARGIN 1.0f

HpBar::HpBar() {
    
}

HpBar::~HpBar() {
    
}

HpBar* HpBar::create( const cocos2d::Size& size ) {
    HpBar* ret = new HpBar();
    if( ret && ret->init( size ) ) {
        ret->autorelease();
        return ret;
    }
    else {
        CC_SAFE_DELETE( ret );
        return nullptr;
    }
}

bool HpBar::init( const cocos2d::Size& size ) {
    if( !Node::init() ) {
        return false;
    }
    
    this->ignoreAnchorPointForPosition( false );
    this->setAnchorPoint( Point( 0.5f, 0.5f ) );
    this->setContentSize( size );
    
    Size inner_size = Size( size.width - 2 * H_MARGIN, size.height - 2 * V_MARGIN );
    
    _background = this->addBar( Color4F( 0, 0, 0, 1.0f ), size, BACKGROUND_TEXTURE_NAME );
    _red_bar = this->addBar( Color4F( 1.0f, 0, 0, 1.0f ), inner_size, RED_TEXTURE_NAME );
    _yellow_bar = this->addBar( Color4F( 1.0f, 1.0f, 0, 1.0f ), inner_size, YELLOW_TEXTURE_NAME );
    _green_bar = this->addBar( Color4F( 0, 1.0f, 0.1843f, 1.0f ), inner_size, GREEN_TEXTUER_NAME );
    
    _percentage = 100.0f;
    this->updateVisible();
    
    return true;
}

void HpBar::setPercentage( float percentage ) {
    if( percentage != _percentage ) {
        _percentage = percentage;
        _yellow_bar->setPercentage( percentage );
        _green_bar->setPercentage( percentage );
        _red_bar->setPercentage( percentage );
        this->updateVisible();
    }
}

float HpBar::getPercentage() {
    return _percentage;
}

void HpBar::updateVisible() {
    if( _percentage <= RED_PERCENTAGE ) {
        _red_bar->setVisible( true );
        _yellow_bar->setVisible( false );
        _green_bar->setVisible( false );
    }
    else if( _percentage <= YELLOW_PERCENTAGE ) {
        _red_bar->setVisible( false );
        _yellow_bar->setVisible( true );
        _green_bar->setVisible( false );
    }
    else {
        _red_bar->setVisible( false );
        _yellow_bar->setVisible( false );
        _green_bar->setVisible( true );
    }
}

//private method
ProgressTimer* HpBar::addBar( const cocos2d::Color4F& color, const cocos2d::Size& size, const std::string& texture_name ) {
    cocos2d::ProgressTimer* bar = nullptr;
    RenderTexture* texture = RenderTexture::create( 1.0f, 1.0f );
    texture->beginWithClear( color.r, color.g, color.b, color.a );
    texture->end();
    Sprite* sp = Sprite::createWithTexture( texture->getSprite()->getTexture() );
    bar = ProgressTimer::create( sp );
    bar->setType( cocos2d::ProgressTimer::Type::BAR );
    bar->setMidpoint( Point( 0, 0.5f ) );
    bar->setBarChangeRate( Point( 1.0f, 0 ) );
    bar->setPercentage( 100.0f );
    bar->setPosition( Point::ZERO );
    bar->setScale( size.width, size.height );
    this->addChild( bar );
    return bar;
}